File: 2021/09/07/relative_ranks.py
"""
https://leetcode.com/problems/relative-ranks/
Author: Andrew Jarombek
Date: 9/7/2021
"""

from typing import List


def find_relative_ranks(score: List[int]) -> List[str]:
    ordered_score = [(s, index) for index, s in enumerate(score)]
    ordered_score.sort(key=lambda x: x[0], reverse=True)

    result = list(range(len(score)))

    for index, (s, i) in enumerate(ordered_score):
        if index == 0:
            val = "Gold Medal"
        elif index == 1:
            val = "Silver Medal"
        elif index == 2:
            val = "Bronze Medal"
        else:
            val = str(index + 1)

        result[i] = val

    return result


if __name__ == '__main__':
    assert find_relative_ranks([5, 4, 3, 2, 1]) == ["Gold Medal", "Silver Medal", "Bronze Medal", "4", "5"]

File: 2022/03/25/relativeRanks.ts
/**
 * https://leetcode.com/problems/relative-ranks/
 * @author Andrew Jarombek
 * @since 3/25/2022
 */

function findRelativeRanks(scores: number[]): string[] {
    let list = scores.map((score, index) => ({score, index}));
    list.sort((a, b) => b.score - a.score);

    let result = scores.map(x => "");
    result[list[0].index] = "Gold Medal";

    if (list.length > 1)
        result[list[1].index] = "Silver Medal";

    if (list.length > 2)
        result[list[2].index] = "Bronze Medal";

    for (let i = 3; i < list.length; i++) {
        result[list[i].index] = `${i+1}`;
    }

    return result;
}